segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/SystemJournalTests.java

```diff
@@ -72,7 +72,12 @@ protected int getThreadPoolSize() {
     }
 
     protected ChunkMetadataStore getMetadataStore() throws Exception {
-        return new InMemoryMetadataStore(ChunkedSegmentStorageConfig.DEFAULT_CONFIG, executorService());
+        val metadataStore = new InMemoryMetadataStore(ChunkedSegmentStorageConfig.DEFAULT_CONFIG, executorService());
```

Comment:
Could you please add as a new test the code provided by @andreipaduroiu that reproduces this issue?

Reply:
It was careful log analysis that revealed that deadlock has occurred. The test doesn't itself detect deadlock and fail.
Because of async nature of flush to LTS it is difficult to write such tests.

However here I'm doing something even better - I already have unit tests that invoke complicated read/write/boot scenarios for system tests. Here I have added extra check to the unit tests on system tests to assert on any attempt to read from store. This will detect any unintentional call to metadata store in case of metadata segments and will fail immediately.

Reply:
@sachin-j-joshi I know what entailed to uncover the issue and the fact that @andreipaduroiu reproduced that locally was critical to uncover it. Not only this, all the existing tests were not enough to detect this problem, for this reason the bug got merged (irrespective if it was a previous bug in combination with a recent test). So, is it too much to ask to add a test (perhaps similar to the code used to reproduce this bug) to have more confidence on the fix?

Reply:
- The dead lock is caused by very complex interactions only possible in fully integrated system.  
- The root cause is re-entrant calls to storage metadata segments during operations on system segments. 
- As a design principle operation on container metadata segment (CMS) and storage metadata segments (SMS) should use only in memory data and should not have any dependency what so ever on any other operation on the same segments. This rule was unintentionally broken. But that did not cause deadlock because we had per segment serialization of requests which was [changed only recently](https://github.com/pravega/pravega/pull/6185/files#diff-e8690fd36836c2a4aa9501f65eddf54431a88ba86c779e3f78f737ab631162d6R756) to serializing on the group
- Solution is to  prevent such reentrant calls. 
- Now - The unit tests make sure no such re-entrant calls are ever made.


Why do you need to have complicated and unreliable tests, when you can check the assumptions of "no circular dependency" on each and every method call on storage system segments?

Yes I think it is difficult to recreate exact end to end test, especially I don't know how best we can detect presence of deadlock and what exactly to assert on? 

Relying on such integration tests to validate does not seem correct approach to me here.- when you can validate more directly We have narrowed down a problem, have a specific fix and then also testing/validating it in component level tests.

Reply:
To be clear there are already several unit tests that create and test reentry/deadlock scenario.
In unit tests by necessity we use **mock implementation of key value store** and as such it doesn't internally call actual openWrite which is a totally expected re-entrant call here. 

Existing tests -

1. https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/SystemJournalTests.java#L1223
2. 
3. https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorageTests.java#L2426
4. 
5. https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorageTests.java#L2558
6. 
7. https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorageTests.java#L2632
8.